### PR TITLE
Add support to query ids returning arrays and objects

### PR DESCRIPTION
**Pull request by HoLo85**
Added capability to store message arrays in custom messages.

### CODE REVIEW

File: sma-webconnect.js [hunk 1]
```diff
@@ -184,15 +184,21 @@ module.exports = function(RED) {
...
+              // iterate over messages from sma device
               for (const key in set) {
-                if ( message.values.hasOwnProperty(key) ) { 
+                if (message.values.hasOwnProperty(key)) {
                   const value = set[key];
-                  
+
                   if (value != null) {
+                    var val_cnt = 0;  // value counter
+                    var val_arr = []; // value array
```
Comment (rlindner) [inline sma-webconnect.js:194]:
```suggestion
                    var values = [];
```

File: sma-webconnect.js [hunk 2]
```diff
@@ -205,9 +211,18 @@ module.exports = function(RED) {
                           }
                         }
 
-                        result[message.values[key].name] = tmp;
+                        // add values to the array
+                        val_arr[val_cnt++] = tmp;
```
Comment (rlindner) [inline sma-webconnect.js:215]:
```suggestion
                        values.push(tmp);
```

File: sma-webconnect.js [hunk 3]
```diff
@@ -205,9 +211,18 @@ module.exports = function(RED) {
...
                       }
                     }
+
+                    // store value in output message
+                    if (val_cnt > 1) {
+                      result[message.values[key].name] = val_arr;
+                    }
+                    else {
+                      result[message.values[key].name] = val_arr[0];
```
Comment (rlindner) [inline sma-webconnect.js:224]:
```suggestion
                    if (values.length > 1) {
                      result[message.values[key].name] = values;
                    } else {
                      result[message.values[key].name] = values[0];
```

**Review by rlindner — CHANGES_REQUESTED** [review]
First of all, thanks again for your contribution and sorry again for the delay :bow:

I wasn't really able to test this properly, but from my understanding the values in question are of the following format?

```json
"6100_40263F00": {
  "7": [
    {
      "val": 123
    },
    {
      "val": 456
    }
  ]
}
```

If that's the case, everything should be OK. Just one nit pick: Please let's just use an array and get rid of that counter.

**Comment by HoLo85** [comment]
hi, updated the PR.
The returned array values look like this with the current implementation:
`{"dc_power":[2067,479],"dc_voltage":[58645,40779],"dc_current":[3527,1178],"available_sessions":2}`

**Comment by HoLo85** [comment]
Had to do an update to the code so messages with objects get stored if requested. The current output will stay the same as now but extended by the object message.

Input:
`{
    "sma_config": {
        "id": "1",
        "values": {
            "6100_0046E500": {
                "name": "phase1_voltage",
                "divider": 1
            },
            "6380_40251E00": {
                "name": "dc_power",
                "divider": 1
            },
            "6380_40451F00": {
                "name": "dc_voltage",
                "divider": 1
            },
            "6380_40452100": {
                "name": "dc_current",
                "divider": 1
            },
            "6180_00522900": {
                "name": "unknown",
                "divider": 1
            },
            "6180_08652500": {
                "name": "unknown2",
                "divider": 1
            }
        }
    }
}`
Output:
`{
    "phase1_voltage":23543,
    "dc_power":[156,404],
    "dc_voltage":[56146,33022],
    "dc_current":[279,1226],
    "unknown":{
        "low":0,
        "high":0,
        "val":null
    },
    "unknown2":{
        "tag":303
    },
    "available_sessions":1
}`

**Pull request merged**